### ISSUE TITLE
[Question] JIT codegen difference between 6.0 and 7.0 for BitOperations.Log2_uint

### ISSUE THREAD

**Issue by alexcovington** [dotnet/runtime#79058]
I'm looking at a similar test case to [this microbenchmark](https://github.com/dotnet/performance/blob/8202bcc76e6032d597050058e75e0f48788e4251/src/benchmarks/micro/libraries/System.Numerics.BitOperations/Perf_BitOperations.cs#L44-L53) and had a few questions about codegen differences between 6.0 and 7.0.

<details>
<summary>C# Test Case</summary>

```
public static int Log2_uint()
{
    int sum = 0;
    uint[] input = input_uint;
    for (int i = 0; i < input.Length; i++)
    {
        sum += BitOperations.Log2(input[i]);
    }
    return sum;
}
```
</details>

<details>
<summary>.NET 6.0 Codegen</summary>

```
G_M56802_IG01:        ; func=00, offs=000000H, size=0000H, bbWeight=1    PerfScore 0.00, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref, nogc <-- Prolog IG
G_M56802_IG02:        ; offs=000000H, size=0020H, bbWeight=1    PerfScore 6.25, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, BB01 [0000], byref, isz, align

IN0001: 000000 xor      eax, eax
IN0002: 000002 mov      rdx, 0x23D18002E90
IN0003: 00000C mov      rdx, gword ptr [rdx]
IN0004: 00000F xor      ecx, ecx
IN0005: 000011 mov      r8d, dword ptr [rdx+8]
IN0006: 000015 test     r8d, r8d
IN0007: 000018 jle      SHORT G_M56802_IG04
IN0008: 00001A align    [6 bytes]

G_M56802_IG03:        ; offs=000020H, size=001FH, bbWeight=4    PerfScore 26.00, gcrefRegs=00000004 {rdx}, byrefRegs=00000000 {}, loop=IG03, BB04 [0012], byref, isz

IN0009: 000020 movsxd   r9, ecx
IN000a: 000023 mov      r9d, dword ptr [rdx+4*r9+16]
IN000b: 000028 or       r9d, 1
IN000c: 00002C lzcnt    r9d, r9d
IN000d: 000031 xor      r9d, 31
IN000e: 000035 add      eax, r9d
IN000f: 000038 inc      ecx
IN0010: 00003A cmp      r8d, ecx
IN0011: 00003D jg       SHORT G_M56802_IG03

G_M56802_IG04:        ; offs=00003FH, size=0001H, bbWeight=1    PerfScore 1.00, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref, epilog, nogc

IN0012: 00003F ret   
```
</details>

<details>
<summary>.NET 7.0 Codegen</summary>

```
G_M56802_IG01:        ; func=00, offs=000000H, size=0000H, bbWeight=1    PerfScore 0.00, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref, nogc <-- Prolog IG
G_M56802_IG02:        ; offs=000000H, size=0020H, bbWeight=1    PerfScore 6.25, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, BB01 [0000], byref, isz, align

IN0001: 000000 xor      eax, eax
IN0002: 000002 mov      rdx, 0x195CDC01EC0      ; data for BitOpsLog2Repro.Program:input_uint
IN0003: 00000C mov      rdx, gword ptr [rdx]
IN0004: 00000F xor      ecx, ecx
IN0005: 000011 mov      r8d, dword ptr [rdx+08H]
IN0006: 000015 test     r8d, r8d
IN0007: 000018 jle      SHORT G_M56802_IG04
IN0008: 00001A align    [6 bytes for IG03]

G_M56802_IG03:        ; offs=000020H, size=001FH, bbWeight=4    PerfScore 26.00, gcrefRegs=00000004 {rdx}, byrefRegs=00000000 {}, loop=IG03, BB02 [0001], byref, isz

IN0009: 000020 mov      r9d, ecx
IN000a: 000023 mov      r9d, dword ptr [rdx+4*r9+10H]
IN000b: 000028 or       r9d, 1
IN000c: 00002C lzcnt    r9d, r9d
IN000d: 000031 xor      r9d, 31
IN000e: 000035 add      eax, r9d
IN000f: 000038 inc      ecx
IN0010: 00003A cmp      r8d, ecx
IN0011: 00003D jg       SHORT G_M56802_IG03

G_M56802_IG04:        ; offs=00003FH, size=0001H, bbWeight=1    PerfScore 1.00, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref, epilog, nogc

IN0012: 00003F ret 
```
</details>

Specifically, there is a difference with how the `GT_CAST` gets compiled. 6.0 uses a `movsxd` and 7.0 uses a `mov`:

**.NET 6.0**:

```
IN0009: 000020 movsxd   r9, ecx
IN000a: 000023 mov      r9d, dword ptr [rdx+4*r9+16]
```

**.NET 7.0**:

```
IN0009: 000020 mov      r9d, ecx
IN000a: 000023 mov      r9d, dword ptr [rdx+4*r9+10H]
```

This looks to be a result of #57970 forcing the cast to be `uint -> long`. I have a few questions about this:
- Is this done just for performance, or were there other reasons for the change? I saw many improvements reported in the #57970 thread, but just wanted to make sure I had the full context of why the change was made.
- The source register (`ecx` in this case) is always guaranteed to be non-negative?

Another question I had was regarding alignment of the loop. When generating a JIT dump, if I set `COMPlus_TieredCompilation=0` and `COMPlus_ReadyToRun=0`, I see a different amount of padding used to align the loop:

<details>
<summary>Codegen</summary>

`COMPlus_TieredCompilation=0`, `COMPlus_ReadyToRun=0`

```
*************** After end code gen, before unwindEmit()
G_M56802_IG01:        ; func=00, offs=000000H, size=0005H, bbWeight=1    PerfScore 1.25, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref, nogc <-- Prolog IG

IN0016: 000000 push     rsi
IN0017: 000001 sub      rsp, 32

G_M56802_IG02:        ; offs=000005H, size=003BH, bbWeight=1    PerfScore 7.75, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, BB01 [0000], byref, isz, align

IN0001: 000005 xor      esi, esi
IN0002: 000007 mov      rcx, 0x7FFF2FD37A30
IN0003: 000011 mov      edx, 4
IN0004: 000016 call     CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
IN0005: 00001B mov      rax, 0x25442001EC0      ; data for BitOpsLog2Repro.Program:input_uint
IN0006: 000025 mov      rax, gword ptr [rax]
IN0007: 000028 xor      edx, edx
IN0008: 00002A mov      ecx, dword ptr [rax+08H]
IN0009: 00002D test     ecx, ecx
IN000a: 00002F jle      SHORT G_M56802_IG04
IN000b: 000031 align    [15 bytes for IG03]

G_M56802_IG03:        ; offs=000040H, size=001EH, bbWeight=4    PerfScore 26.00, gcrefRegs=00000001 {rax}, byrefRegs=00000000 {}, loop=IG03, BB02 [0001], byref, isz

IN000c: 000040 mov      r8d, edx
IN000d: 000043 mov      r8d, dword ptr [rax+4*r8+10H]
IN000e: 000048 or       r8d, 1
IN000f: 00004C lzcnt    r8d, r8d
IN0010: 000051 xor      r8d, 31
IN0011: 000055 add      esi, r8d
IN0012: 000058 inc      edx
IN0013: 00005A cmp      ecx, edx
IN0014: 00005C jg       SHORT G_M56802_IG03

G_M56802_IG04:        ; offs=00005EH, size=0002H, bbWeight=1    PerfScore 0.25, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, BB03 [0003], byref

IN0015: 00005E mov      eax, esi

G_M56802_IG05:        ; offs=000060H, size=0006H, bbWeight=1    PerfScore 1.75, epilog, nogc, extend

IN0018: 000060 add      rsp, 32
IN0019: 000064 pop      rsi
IN001a: 000065 ret   
```

`COMPlus_TieredCompilation=1`, `COMPlus_ReadyToRun=1`

```
*************** After end code gen, before unwindEmit()
G_M56802_IG01:        ; func=00, offs=000000H, size=0000H, bbWeight=1    PerfScore 0.00, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref, nogc <-- Prolog IG
G_M56802_IG02:        ; offs=000000H, size=0020H, bbWeight=1    PerfScore 6.25, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, BB01 [0000], byref, isz, align

IN0001: 000000 xor      eax, eax
IN0002: 000002 mov      rdx, 0x25C9DC01EC0      ; data for BitOpsLog2Repro.Program:input_uint
IN0003: 00000C mov      rdx, gword ptr [rdx]
IN0004: 00000F xor      ecx, ecx
IN0005: 000011 mov      r8d, dword ptr [rdx+08H]
IN0006: 000015 test     r8d, r8d
IN0007: 000018 jle      SHORT G_M56802_IG04
IN0008: 00001A align    [6 bytes for IG03]

G_M56802_IG03:        ; offs=000020H, size=001FH, bbWeight=4    PerfScore 26.00, gcrefRegs=00000004 {rdx}, byrefRegs=00000000 {}, loop=IG03, BB02 [0001], byref, isz

IN0009: 000020 mov      r9d, ecx
IN000a: 000023 mov      r9d, dword ptr [rdx+4*r9+10H]
IN000b: 000028 or       r9d, 1
IN000c: 00002C lzcnt    r9d, r9d
IN000d: 000031 xor      r9d, 31
IN000e: 000035 add      eax, r9d
IN000f: 000038 inc      ecx
IN0010: 00003A cmp      r8d, ecx
IN0011: 00003D jg       SHORT G_M56802_IG03

G_M56802_IG04:        ; offs=00003FH, size=0001H, bbWeight=1    PerfScore 1.00, gcrefRegs=00000000 {}, byrefRegs=00000000 {}, byref, epilog, nogc

IN0012: 00003F ret      
```
</details>

Is the difference in padding only due the extra call before the loop? Or is something else affecting the alignment in this case?

Please let me know if I can clarify any of the examples/output above. Thanks so much!

**Comment by JulieLeeMSFT** [comment]
cc @EgorBo @tannergooding, PTAL.

**Comment by tannergooding** [comment]
> Specifically, there is a difference with how the GT_CAST gets compiled. 6.0 uses a movsxd and 7.0 uses a mov:

Normally `signed values` (such as `int`) are sign extended where-as `unsigned values` (such as `uint`) are zero extended.

Sometime in .NET 7 we added handling to better understand cases where a `signed value` was considered "never negative" which allows us to emit a zero-extension instead.

> Another question I had was regarding alignment of the loop. When generating a JIT dump, if I set COMPlus_TieredCompilation=0 and COMPlus_ReadyToRun=0, I see a different amount of padding used to align the loop:

There are many factors that can impact loop alignment and the heuristics have been changed a few times. We generally try to determine that alignment is "profitable" and only align loops where we believe that to be the case.
